File: holoken-core/src/main/java/com/holokenmod/creation/GridBuilder.java
package com.holokenmod.creation;

import com.holokenmod.grid.Grid;
import com.holokenmod.grid.GridCage;
import com.holokenmod.grid.GridCageAction;
import com.holokenmod.grid.GridSize;
import com.holokenmod.options.DigitSetting;
import com.holokenmod.options.GameOptionsVariant;
import com.holokenmod.options.GameVariant;

import java.util.ArrayList;
import java.util.Arrays;

public class GridBuilder {
	private final Grid grid;
	private int cageId = 0;
	private ArrayList<Integer> values = new ArrayList<>();
	
	public GridBuilder(int size) {
		this(size, size);
	}
	
	public GridBuilder(int size, DigitSetting digitSetting) {
		this(size, size, GameOptionsVariant.createClassic(digitSetting));
	}
	
	public GridBuilder(int size, GameOptionsVariant variant) {
		this(size, size, variant);
	}
	
	public GridBuilder(int width, int heigth) {
		this(width, heigth, GameOptionsVariant.createClassic());
	}
	
	public GridBuilder(int width, int heigth, GameOptionsVariant variant) {
		grid = new Grid(new GameVariant(new GridSize(width, heigth), variant));
		
		grid.addAllCells();
	}
	
	public GridBuilder addCage(int result, GridCageAction action, int... cellIds) {
		if ( cellIds == null || cellIds.length == 0) {
			throw new RuntimeException("No cell ids given.");
		}
		
		GridCage cage = new GridCage(grid);
		
		cage.setCageId(cageId++);
		cage.setAction(action);
		cage.setResult(result);
		
		for(int cellId : cellIds) {
			cage.addCell(grid.getCell(cellId));
		}
		
		grid.addCage(cage);

		return this;
	}
	
	public void addValueRow(Integer... values) {
		this.values.addAll(Arrays.asList(values));
	}

	public Grid createGrid() {
		grid.setCageTexts();
		
		if (!values.isEmpty()) {
			int cellId = 0;
			
			for(int value: values) {
				grid.getCell(cellId).setValue(value);
				
				cellId++;
			}
		}
		
		return grid;
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/creation/GridBuilder.kt
package com.holokenmod.creation

import com.holokenmod.creation.cage.GridCageType
import com.holokenmod.grid.Grid
import com.holokenmod.grid.GridCage
import com.holokenmod.grid.GridCageAction
import com.holokenmod.grid.GridSize
import com.holokenmod.options.DigitSetting
import com.holokenmod.options.GameOptionsVariant
import com.holokenmod.options.GameVariant

class GridBuilder constructor(
    width: Int,
    heigth: Int,
    variant: GameOptionsVariant = GameOptionsVariant.createClassic()
) {
    private var cages = mutableListOf<GridCage>()
    private val grid: Grid
    private var cageId = 0
    private val values = mutableListOf<Int>()

    constructor(size: Int) : this(size, size)
    constructor(size: Int, digitSetting: DigitSetting) : this(
        size,
        size,
        GameOptionsVariant.createClassic(digitSetting)
    )

    constructor(width: Int, height: Int, digitSetting: DigitSetting) : this(
        width,
        height,
        GameOptionsVariant.createClassic(digitSetting)
    )

    init {
        grid = Grid(GameVariant(GridSize(width, heigth), variant))
    }

    fun addSingleCage(result: Int, cellId: Int): GridBuilder {
        return addCage(result, GridCageAction.ACTION_NONE, GridCageType.SINGLE, cellId)
    }

    fun addCage(result: Int, action: GridCageAction, cageType: GridCageType, firstCellId: Int): GridBuilder {
        val firstCell = grid.getCell(firstCellId)

        val cage = GridCage.createWithCells(cageId++, grid, action, firstCell, cageType)
        cage.result = result

        cages += cage

        return this
    }

    fun addValueRow(vararg values: Int) {
        this.values += values.toList()
    }

    fun createGrid(): Grid {
        grid.setCageTexts()

        cages.forEach { grid.addCage(it) }

        if (values.isNotEmpty()) {
            var cellId = 0
            values.forEach {
                grid.getCell(cellId).value = it
                cellId++
            }
        }
        return grid
    }
}
